File: crates/psychevo-gateway/tests/fixtures/fake_acp_session_persistence.py
#!/usr/bin/env python3
import json
import os
import sqlite3
import sys

loaded_session = None
counter_path = sys.argv[0] + ".counter"
process_counter_path = sys.argv[0] + ".processes"
try:
    with open(process_counter_path, "r", encoding="utf-8") as process_counter_file:
        process_counter = int(process_counter_file.read())
except (FileNotFoundError, ValueError):
    process_counter = 0
with open(process_counter_path, "w", encoding="utf-8") as process_counter_file:
    process_counter_file.write(str(process_counter + 1))


def send(value):
    print(json.dumps(value), flush=True)


def update(session_id, update):
    send({"jsonrpc": "2.0", "method": "session/update", "params": {
        "sessionId": session_id,
        "update": update
    }})


for line in sys.stdin:
    if not line.strip():
        continue
    message = json.loads(line)
    method = message.get("method")
    mid = message.get("id")
    params = message.get("params") or {}
    if method == "initialize":
        send({"jsonrpc": "2.0", "id": mid, "result": {"protocolVersion": 1, "agentCapabilities": {}}})
    elif method == "session/new":
        try:
            with open(counter_path, "r", encoding="utf-8") as counter_file:
                counter = int(counter_file.read())
        except (FileNotFoundError, ValueError):
            counter = 0
        counter += 1
        with open(counter_path, "w", encoding="utf-8") as counter_file:
            counter_file.write(str(counter))
        send({"jsonrpc": "2.0", "id": mid, "result": {"sessionId": "native-" + str(counter)}})
    elif method == "session/load":
        loaded_session = params.get("sessionId")
        update(loaded_session, {
            "sessionUpdate": "agent_message_chunk",
            "content": {"type": "text", "text": "old answer from loaded history"}
        })
        send({"jsonrpc": "2.0", "id": mid, "result": {}})
    elif method == "session/prompt":
        session_id = params.get("sessionId") or "native-1"
        with sqlite3.connect(os.environ["PSYCHEVO_BINDING_DB"]) as connection:
            persisted = connection.execute(
                "SELECT native_session_id FROM gateway_runtime_bindings WHERE native_session_id = ?",
                (session_id,),
            ).fetchone()
        if persisted != (session_id,):
            raise RuntimeError("native session binding was not persisted before prompt")
        chunks = []
        for block in params.get("prompt") or []:
            if block.get("type") == "text":
                chunks.append(block.get("text") or "")
        prefix = "loaded:" + loaded_session if loaded_session else "new:" + session_id
        text = prefix + ":" + "\n".join(chunks)
        send({"jsonrpc": "2.0", "method": "session/update", "params": {
            "sessionId": session_id,
            "update": {
                "sessionUpdate": "agent_message_chunk",
                "content": {"type": "text", "text": text}
            }
        }})
        send({"jsonrpc": "2.0", "id": mid, "result": {"stopReason": "end_turn"}})
    else:
        send({"jsonrpc": "2.0", "id": mid, "error": {"code": -32601, "message": "method not found"}})

File: crates/psychevo-gateway/tests/fixtures/fake_acp_session_persistence.js
#!/usr/bin/env node
const fs = require("node:fs");
const readline = require("node:readline");
const { DatabaseSync } = require("node:sqlite");
let loadedSession = null;
const counterPath = `${__filename}.counter`;
const processCounterPath = `${__filename}.processes`;
let processCounter = 0;
try { processCounter = Number.parseInt(fs.readFileSync(processCounterPath, "utf8"), 10) || 0; } catch {}
fs.writeFileSync(processCounterPath, String(processCounter + 1), "utf8");
function send(value) { process.stdout.write(`${JSON.stringify(value)}\n`); }
function update(sessionId, value) { send({ jsonrpc: "2.0", method: "session/update", params: { sessionId, update: value } }); }
function bindingExists(sessionId) {
  const db = new DatabaseSync(process.env.PSYCHEVO_BINDING_DB, { readOnly: true });
  try {
    return db.prepare("SELECT native_session_id FROM gateway_runtime_bindings WHERE native_session_id = ?").get(sessionId)?.native_session_id === sessionId;
  } finally { db.close(); }
}
function handle(message) {
  const { method, id } = message; const params = message.params || {};
  if (method === "initialize") send({ jsonrpc: "2.0", id, result: { protocolVersion: 1, agentCapabilities: {} } });
  else if (method === "session/new") {
    let counter = 0; try { counter = Number.parseInt(fs.readFileSync(counterPath, "utf8"), 10) || 0; } catch {}
    counter += 1; fs.writeFileSync(counterPath, String(counter), "utf8");
    send({ jsonrpc: "2.0", id, result: { sessionId: `native-${counter}` } });
  } else if (method === "session/load") {
    loadedSession = params.sessionId;
    update(loadedSession, { sessionUpdate: "agent_message_chunk", content: { type: "text", text: "old answer from loaded history" } });
    send({ jsonrpc: "2.0", id, result: {} });
  } else if (method === "session/prompt") {
    const sessionId = params.sessionId || "native-1";
    if (!bindingExists(sessionId)) throw new Error("native session binding was not persisted before prompt");
    const chunks = (params.prompt || []).filter((block) => block.type === "text").map((block) => block.text || "");
    const prefix = loadedSession ? `loaded:${loadedSession}` : `new:${sessionId}`;
    update(sessionId, { sessionUpdate: "agent_message_chunk", content: { type: "text", text: `${prefix}:${chunks.join("\n")}` } });
    send({ jsonrpc: "2.0", id, result: { stopReason: "end_turn" } });
  } else send({ jsonrpc: "2.0", id, error: { code: -32601, message: "method not found" } });
}
readline.createInterface({ input: process.stdin, crlfDelay: Infinity }).on("line", (line) => { if (line.trim()) handle(JSON.parse(line)); });
